File: src/200.number-of-islands/solution_bfs.cpp
/*
 * @lc app=leetcode.cn id=200 lang=cpp
 *
 * https://leetcode-cn.com/problems/number-of-islands/
 */

#include <queue>
#include <utility>
#include <vector>

using namespace std;

// @lc code=start

class Solution {
 private:
  // grid is not empty
  // row, col are valid
  void resetIslandBfs(vector<vector<char>> &grid, int row, int col) {
    int row_num = grid.size();
    int col_num = grid[0].size();

    // a queue to store unhandled elements of the island
    queue<pair<int, int>> elms;

    // reset grid element and push it to queue
    grid[row][col] = '0';
    elms.push({row, col});

    while (!elms.empty()) {
      pair<int, int> rc = elms.front();
      elms.pop();
      int r = rc.first;
      int c = rc.second;

      // up
      if (r - 1 >= 0 && grid[r - 1][c] == '1') {
        grid[r - 1][c] = '0';
        elms.push({r - 1, c});
      }
      // down
      if (r + 1 < row_num && grid[r + 1][c] == '1') {
        grid[r + 1][c] = '0';
        elms.push({r + 1, c});
      }
      // left
      if (c - 1 >= 0 && grid[r][c - 1] == '1') {
        grid[r][c - 1] = '0';
        elms.push({r, c - 1});
      }
      // right
      if (c + 1 < col_num && grid[r][c + 1] == '1') {
        grid[r][c + 1] = '0';
        elms.push({r, c + 1});
      }
    }
  }

 public:
  int numIslands(vector<vector<char>> &grid) {
    // grid is empty
    if (grid.size() == 0 || grid[0].size() == 0) {
      return 0;
    }

    int row_num = grid.size();
    int col_num = grid[0].size();
    int island_num = 0;

    for (int row = 0; row < row_num; ++row) {
      for (int col = 0; col < col_num; ++col) {
        if (grid[row][col] == '1') {
          island_num += 1;
          resetIslandBfs(grid, row, col);
        }
      }
    }

    return island_num;
  }
};

// @lc code=end

File: src/200.number-of-islands/solution_bfs.ts
/*
 * @lc app=leetcode.cn id=200 lang=typescript
 *
 * https://leetcode-cn.com/problems/number-of-islands/
 */

// @lc code=start

// grid is not empty
// row, col are valid
function clearIslandBfs(grid: string[][], row: number, col: number) {
  const rowNum = grid.length;
  const colNum = grid[0].length;

  // a queue to store unhandled elements of the island
  // queue element: [rowIndex, colIndex]
  const queue: [number, number][] = [];

  // clear grid element and push it to queue
  grid[row][col] = '0';
  queue.push([row, col]);

  while (queue.length > 0) {
    const rc = queue.shift() as [number, number];
    const r = rc[0];
    const c = rc[1];

    // up
    if (r - 1 >= 0 && grid[r - 1][c] === '1') {
      grid[r - 1][c] = '0';
      queue.push([r - 1, c]);
    }
    // down
    if (r + 1 < rowNum && grid[r + 1][c] === '1') {
      grid[r + 1][c] = '0';
      queue.push([r + 1, c]);
    }
    // left
    if (c - 1 >= 0 && grid[r][c - 1] === '1') {
      grid[r][c - 1] = '0';
      queue.push([r, c - 1]);
    }
    // right
    if (c + 1 < colNum && grid[r][c + 1] === '1') {
      grid[r][c + 1] = '0';
      queue.push([r, c + 1]);
    }
  }
}

function numIslands(grid: string[][]): number {
  // grid is empty
  if (grid.length === 0 || grid[0].length === 0) {
    return 0;
  }

  const rowNum = grid.length;
  const colNum = grid[0].length;
  let islandNum = 0;

  for (let row = 0; row < rowNum; ++row) {
    for (let col = 0; col < colNum; ++col) {
      if (grid[row][col] === '1') {
        islandNum += 1;
        clearIslandBfs(grid, row, col);
      }
    }
  }

  return islandNum;
}

// @lc code=end

export default numIslands;
